src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java

```diff
@@ -108,6 +109,9 @@
     private final WebSocketInputStateMachine inputStateMachine;
     private final WebSocketOutputStateMachine outputStateMachine;
     private final WebSocketExtensionFactory extensionFactory;
+    private final OptimisticReentrantLock readLock;
+    private final OptimisticReentrantLock stateLock;
+    private final OptimisticReentrantLock writeLock;
 
     private String negotiatedProtocol;
     private WsInputStream inputStream;
```

Comment:
Since the `happens-before` relationship will kick-in only after the second thread acquires the lock released by the first thread, I think you are right.